lib/PuppeteerSharp/Helpers/TaskQueue.cs

```diff
@@ -10,24 +10,27 @@ internal sealed class TaskQueue : IDisposable, IAsyncDisposable
         [SuppressMessage("Usage", "CA2213: Disposable fields should be disposed", Justification = "The disposable field is being disposed asynchronously.")]
         [SuppressMessage("CodeQuality", "IDE0079:Remove unnecessary suppression", Justification = "The CA2213 suppression is actually necessary.")]
         private readonly SemaphoreSlim _semaphore;
-        private bool _isDisposed;
+        private int _disposed;
 
         internal TaskQueue() => _semaphore = new SemaphoreSlim(1);
 
         public void Dispose()
         {
-            if (_isDisposed)
+            if (Interlocked.Exchange(ref _disposed, 1) != 0)
             {
                 return;
             }
 
             _ = Task.Run(() => DisposeAsync());
```

Comment:
I haven't had a chance to test this out, but won't we we immediately return from `DisposeAsync` as `_disposed` has already been changed?

Reply:
I had some time to test, and `_semaphore` is not disposed when calling `Dispose` for the reason described above.

If we decouple `Dispose` and `DisposeAsync` completely and call `SemaphoreSlim.Wait()` in `Dispose`, here are some unit tests to verify it.
With the thread safe guards, we should in addition be able to remove the try/catch of `ObjectDisposedException`.

```cs
public void Dispose()
{
    if (Interlocked.Exchange(ref _disposed, 1) != 0)
    {
        return;
    }

    _semaphore.Wait();
    _semaphore.Dispose();
}

public async ValueTask DisposeAsync()
{
    if (Interlocked.Exchange(ref _disposed, 1) != 0)
    {
        return;
    }

    await _semaphore.WaitAsync().ConfigureAwait(false);
    _semaphore.Dispose();
}
```

```cs
public class TaskQueueTests
{
    [Fact]
    public void ShouldDisposeSemaphoreWhenDisposing()
    {
        var taskQueue = new TaskQueue();
        taskQueue.Dispose();

        var semaphore = GetSemaphore(taskQueue);
        Assert.Throws<ObjectDisposedException>(() => semaphore.AvailableWaitHandle);
    }

    [Fact]
    public void ShouldNotThrowWhenDisposingMultipleTimes()
    {
        var taskQueue = new TaskQueue();
        taskQueue.Dispose();

        // Can be disposed a second time without throwing
        taskQueue.Dispose();
    }

    [Fact]
    public async Task ShouldDisposeSemaphoreWhenDisposingAsync()
    {
        var taskQueue = new TaskQueue();
        await taskQueue.DisposeAsync().ConfigureAwait(false);

        var semaphore = GetSemaphore(taskQueue);
        Assert.Throws<ObjectDisposedException>(() => semaphore.AvailableWaitHandle);
    }

    [Fact]
    public async Task ShouldNotThrowWhenDisposingMultipleTimesAsync()
    {
        var taskQueue = new TaskQueue();
        await taskQueue.DisposeAsync().ConfigureAwait(false);

        // Can be disposed a second time without throwing
        await taskQueue.DisposeAsync().ConfigureAwait(false);
    }

    private static SemaphoreSlim GetSemaphore(TaskQueue queue) =>
        (SemaphoreSlim)typeof(TaskQueue).GetField("_semaphore", BindingFlags.Instance | BindingFlags.NonPublic).GetValue(queue);
}
```

Reply:
> I haven't had a chance to test this out, but won't we we immediately return from DisposeAsync as _disposed has already been changed?

You are absolutely correct! My mistake!